src/interpreter.c

```diff
@@ -20,8 +20,98 @@ typedef struct {
     jl_value_t **locals; // slots for holding local slots and ssavalues
     jl_svec_t *sparam_vals; // method static parameters, if eval-ing a method body
     int preevaluation; // use special rules for pre-evaluating expressions
+    size_t ip; // Points to the currently-evaluating statement
 } interpreter_state;
 
+// Backtrace support;
+#ifdef _OS_LINUX_
+extern uintptr_t __start_jl_enter_interpreter_frame;
+extern uintptr_t __stop_jl_enter_interpreter_frame;
+
+extern uintptr_t __start_jl_interpreter_frame;
+extern uintptr_t __stop_jl_interpreter_frame;
+
+#define SECT_INTERP JL_SECTION("jl_interpreter_frame")
+#define SECT_ENTER_INTERP JL_SECTION("jl_enter_interpreter_frame")
+
+#elif defined(_OS_DARWIN_)
+extern uintptr_t __start_jl_enter_interpreter_frame_val __asm("section$start$__TEXT$__jeif");
+uintptr_t __start_jl_enter_interpreter_frame = (uintptr_t)&__start_jl_enter_interpreter_frame_val;
+extern uintptr_t __stop_jl_enter_interpreter_frame_val __asm("section$end$__TEXT$__jeif");
+uintptr_t __stop_jl_enter_interpreter_frame = (uintptr_t)&__stop_jl_enter_interpreter_frame_val;
+
+extern uintptr_t __start_jl_interpreter_frame_val __asm("section$start$__TEXT$__jif");
+uintptr_t __start_jl_interpreter_frame = (uintptr_t)&__start_jl_interpreter_frame_val;
+extern uintptr_t __stop_jl_interpreter_frame_val __asm("section$end$__TEXT$__jif");
+uintptr_t __stop_jl_interpreter_frame = (uintptr_t)&__stop_jl_interpreter_frame_val;
+
+#define SECT_INTERP JL_SECTION("__TEXT,__jif")
+#define SECT_ENTER_INTERP JL_SECTION("__TEXT,__jeif")
+#else
+JL_DLLEXPORT int jl_is_interpreter_frame(uintptr_t ip)
+{
+    return 0;
+}
+
+JL_DLLEXPORT int jl_is_enter_interpreter_frame(uintptr_t ip)
+{
+    return 0;
+}
+
+#define SECT_INTERP
+#define SECT_ENTER_INTERP
+#define NO_INTERP_BT
+#endif
+
+#ifndef NO_INTERP_BT
+JL_DLLEXPORT int jl_is_interpreter_frame(uintptr_t ip)
+{
+    return __start_jl_interpreter_frame <= ip && ip <= __stop_jl_interpreter_frame;
+}
+
+JL_DLLEXPORT int jl_is_enter_interpreter_frame(uintptr_t ip)
+{
+    return __start_jl_enter_interpreter_frame <= ip && ip <= __stop_jl_enter_interpreter_frame;
+}
+
+JL_DLLEXPORT size_t jl_capture_interp_frame(uintptr_t *data, uintptr_t sp, size_t space_remaining)
+{
+    interpreter_state *s = (interpreter_state *)sp;
```

Comment:
Just checked that this may not work all the time. This is actually architecture dependent and `AArch64` puts frame pointer and link register below (lower address) local variables. Stack protector also changes this on aarch64 though it seems to not affect x86....

Anyway, I think an asm version will be much better. I'll write the arm and aarch64 version when you have the x86 version.

Reply:
Actually it seems that on aarch64, gcc follows the [aapcs64](http://infocenter.arm.com/help/topic/com.arm.doc.ihi0055b/IHI0055B_aapcs64.pdf) recommendation and put local variable above fp and lr while clang/llvm 5.0 put the fp and lr above local variables. Both put stack protection slot at the bottom though..